Handle Content-Type in GET requests

There is a [WAF bypass](https://www.youtube.com/watch?v=tSf_IXfuzXk) ([slides](https://www.slideshare.net/SoroushDalili/waf-bypass-techniques-using-http-standard-and-web-servers-behaviour)) technique to send GET requests with attack payload and Content-Type header - some web servers still process the body of GET requests. In `__req_parse_content_type()` we just skip Content-Type parsing and seems just normally forward such requests to a backend.

Need to check RFC for Content-Type requirements - if it prohibits such requests, then we must block them. Otherwise probably we need a new configuration option like `http_strict_mode` which allows to set Tempesta mode for various  HTTP fields in which it blocks such HTTP messages. If the mode isn't set for a HTTP field, then we should just print warning. There are potentially many HTTP fields which behavior are either not clearly defined by RFC or impractical (e.g. single LF character for end of HTTP line).

Please update [HTTP security](https://github.com/tempesta-tech/tempesta/wiki/HTTP-security) and [Web security](https://github.com/tempesta-tech/tempesta/wiki/Web-security) wikis with detailed examples of the attacks (with reference to the original talk) and Tempesta settings.

The unit test for HTTP parser had a bug fixed in https://github.com/tempesta-tech/tempesta/commit/4c9d287c5e999d807048724459d7da7ca5bca2b6#diff-b1785da4f54bd40909398a1a64bcdeef768a8a30c0764f23cb706d2102cfc657L821 . After the fix many failures on Content-Type validation appeared in the test:
```
[   57.529869] tfw_test: /root/tempesta/fw/t/unit/test_http_parser.c:861: error:
[   57.531219] tfw_test: FAIL:
[   57.531687] tfw_test:   test__http_parser__fills_hdr_tbl_for_req()
[   57.532610] tfw_test:   EXPECT_TRUE((tfw_str_eq_cstr(&h_conttype, s_ct, strlen(s_ct), 0))) => 0x0
[   57.534012] tfw_test: /root/tempesta/fw/t/unit/test_http_parser.c:861: error:
[   57.535564] tfw_test: FAIL:
[   57.536098] tfw_test:   test__http_parser__fills_hdr_tbl_for_req()
[   57.537008] tfw_test:   EXPECT_TRUE((tfw_str_eq_cstr(&h_conttype, s_ct, strlen(s_ct), 0))) => 0x0
[   57.538460] tfw_test: /root/tempesta/fw/t/unit/test_http_parser.c:861: error:
[   57.539565] tfw_test: FAIL:
[   57.540044] tfw_test:   test__http_parser__fills_hdr_tbl_for_req()
[   57.540951] tfw_test:   EXPECT_TRUE((tfw_str_eq_cstr(&h_conttype, s_ct, strlen(s_ct), 0))) => 0x0
[   57.543131] tfw_test: /root/tempesta/fw/t/unit/test_http_parser.c:861: error:
....
```

Besides the fix, there is also the bug with the loop in `split_and_parse_n()` (I belive it's already discussed in other issues/pull requests), which must be fixed.

--- a/fw/http.c
+++ b/fw/http.c
@@ -5495,7 +5495,7 @@ next_msg:
 			&& !TFW_HTTP_IS_METH_SAFE(req->method_override))
 		{
 			tfw_http_req_parse_block(req, 400,
-				"request dropped: unsafe method override attempted");
+				"request dropped: unsafe method override");
 			return TFW_BLOCK;
 		}
 		req->method = req->method_override;

--- a/fw/http.c
+++ b/fw/http.c
@@ -4010,7 +4010,7 @@ tfw_http_hdr_split(TfwStr *hdr, TfwStr *name_out, TfwStr *val_out, bool inplace)
 
 		val_out->len += chunk->len;
 
-		/* 
+		/*
 		 * Skip OWS after the header value (RWS) - they must be in
 		 * separate chunks too.
 		 */
@@ -5485,14 +5485,15 @@ next_msg:
 	 * We don't rewrite the method string and don't remove override header
 	 * since there can be additional intermediates between TempestaFW and
 	 * backend.
-	 * 
+	 *
 	 * While non-idempotent method can be hidden behind idempotent, it is
 	 * reasonable to expect that non-safe method can not be hidden behind
 	 * safe method.
 	 */
 	if (unlikely(req->method_override)) {
 		if (TFW_HTTP_IS_METH_SAFE(req->method)
-			&& !TFW_HTTP_IS_METH_SAFE(req->method_override)) {
+			&& !TFW_HTTP_IS_METH_SAFE(req->method_override))
+		{
 			tfw_http_req_parse_block(req, 400,
 				"request dropped: unsafe method override attempted");
 			return TFW_BLOCK;

--- a/fw/http.c
+++ b/fw/http.c
@@ -5485,9 +5485,21 @@ next_msg:
 	 * We don't rewrite the method string and don't remove override header
 	 * since there can be additional intermediates between TempestaFW and
 	 * backend.
+	 * 
+	 * While non-idempotent method can be hidden behind idempotent, it is
+	 * reasonable to expect that safe method will be hidden only behind
+	 * safe method (usually GET), and non-safe behind non-safe method
+	 * (usually POST).
 	 */
-	if (unlikely(req->method_override))
+	if (unlikely(req->method_override)) {
+		if (TFW_HTTP_IS_METH_SAFE(req->method)
+			&& !TFW_HTTP_IS_METH_SAFE(req->method_override)) {
+			tfw_http_req_parse_block(req, 400,
+				"request dropped: unsafe method override attempted");
+			return TFW_BLOCK;
+		}
 		req->method = req->method_override;
+	}
 
 	if (!TFW_MSG_H2(req))
 		hmsib = tfw_h1_req_process(stream, skb);

--- a/fw/http.c
+++ b/fw/http.c
@@ -5487,9 +5487,8 @@ next_msg:
 	 * backend.
 	 * 
 	 * While non-idempotent method can be hidden behind idempotent, it is
-	 * reasonable to expect that safe method will be hidden only behind
-	 * safe method (usually GET), and non-safe behind non-safe method
-	 * (usually POST).
+	 * reasonable to expect that non-safe method can not be hidden behind
+	 * safe method.
 	 */
 	if (unlikely(req->method_override)) {
 		if (TFW_HTTP_IS_METH_SAFE(req->method)

--- a/fw/http.h
+++ b/fw/http.h
@@ -116,9 +116,9 @@ typedef enum {
 	_TFW_HTTP_METH_COUNT
 } tfw_http_meth_t;
 
-#define TFW_HTTP_IS_METH_SAFE(meth)	\
-(  meth == TFW_HTTP_METH_GET     || meth == TFW_HTTP_METH_HEAD	\
-|| meth == TFW_HTTP_METH_OPTIONS || meth == TFW_HTTP_METH_PROPFIND)
+#define TFW_HTTP_IS_METH_SAFE(meth)									\
+	((meth) == TFW_HTTP_METH_GET || (meth) == TFW_HTTP_METH_HEAD	\
+	 || (meth) == TFW_HTTP_METH_OPTIONS || (meth) == TFW_HTTP_METH_PROPFIND)
 
 /* HTTP protocol versions. */
 enum {

--- a/fw/http.h
+++ b/fw/http.h
@@ -86,6 +86,7 @@ enum {
 };
 
 /* TODO: When CONNECT will be added, add it to tfw_handle_validation_req() */
+/* New safe methods MUST be added to TFW_HTTP_IS_METH_SAFE macro */
 typedef enum {
 	_TFW_HTTP_METH_NONE,
 	/*
@@ -113,6 +114,10 @@ typedef enum {
 	_TFW_HTTP_METH_UNKNOWN,
 	_TFW_HTTP_METH_COUNT
 } tfw_http_meth_t;
+
+#define TFW_HTTP_IS_METH_SAFE(meth)	\
+(  meth == TFW_HTTP_METH_GET     || meth == TFW_HTTP_METH_HEAD	\
+|| meth == TFW_HTTP_METH_OPTIONS || meth == TFW_HTTP_METH_PROPFIND)
 
 /* HTTP protocol versions. */
 enum {

--- a/fw/http.h
+++ b/fw/http.h
@@ -116,7 +116,7 @@ typedef enum {
 	_TFW_HTTP_METH_COUNT
 } tfw_http_meth_t;
 
-#define TFW_HTTP_IS_METH_SAFE(meth)									\
+#define TFW_HTTP_IS_METH_SAFE(meth)					\
 	((meth) == TFW_HTTP_METH_GET || (meth) == TFW_HTTP_METH_HEAD	\
 	 || (meth) == TFW_HTTP_METH_OPTIONS || (meth) == TFW_HTTP_METH_PROPFIND)
 

--- a/fw/http.h
+++ b/fw/http.h
@@ -85,7 +85,8 @@ enum {
 	TFW_HTTP_FSM_DONE	= TFW_GFSM_HTTP_STATE(TFW_GFSM_STATE_LAST)
 };
 
-/* TODO: When CONNECT will be added, add it to tfw_handle_validation_req() */
+/* TODO: When CONNECT will be added, add it to tfw_handle_validation_req()
+ * and to body processing content-length filtering */
 /* New safe methods MUST be added to TFW_HTTP_IS_METH_SAFE macro */
 typedef enum {
 	_TFW_HTTP_METH_NONE,

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1170,9 +1170,9 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 			 || req->method_override == TFW_HTTP_METH_TRACE)	\
 				TFW_PARSER_BLOCK(RGen_BodyInit);			\
 		} else if (req->method == TFW_HTTP_METH_GET	\
-			  || req->method == TFW_HTTP_METH_HEAD	\
-			  || req->method == TFW_HTTP_METH_DELETE	\
-			  || req->method == TFW_HTTP_METH_TRACE) {	\
+				|| req->method == TFW_HTTP_METH_HEAD	\
+				|| req->method == TFW_HTTP_METH_DELETE	\
+				|| req->method == TFW_HTTP_METH_TRACE) {	\
 			TFW_PARSER_BLOCK(RGen_BodyInit);			\
 		}	\
 	}	\

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1156,26 +1156,30 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
 	/* According to RFC 7231 4.3.* a payload within GET, HEAD, DELETE,	\
-	 * TRACE and CONNECT requests has no defined semantics and	\
+	 * TRACE and CONNECT requests has no defined semantics and			\
 	 * implementations can reject it. We do this respecting overrides.	\
 	*/	\
-	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])	\
-	 || !TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE])) {	\
+	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])		\
+		|| !TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE])) {	\
 		/* Method override either honored or request message with method	\
-		 * override header dropped later in processing */	\
-		if (unlikely(req->method_override)) {	\
-			if (req->method_override == TFW_HTTP_METH_GET	\
-			 || req->method_override == TFW_HTTP_METH_HEAD	\
-			 || req->method_override == TFW_HTTP_METH_DELETE	\
-			 || req->method_override == TFW_HTTP_METH_TRACE)	\
-				TFW_PARSER_BLOCK(RGen_BodyInit);			\
-		} else if (req->method == TFW_HTTP_METH_GET	\
-				|| req->method == TFW_HTTP_METH_HEAD	\
-				|| req->method == TFW_HTTP_METH_DELETE	\
-				|| req->method == TFW_HTTP_METH_TRACE) {	\
-			TFW_PARSER_BLOCK(RGen_BodyInit);			\
-		}	\
-	}	\
+		 * override header dropped later in processing */					\
+		if (unlikely(req->method_override)) {					\
+			if (req->method_override == TFW_HTTP_METH_GET		\
+				|| req->method_override == TFW_HTTP_METH_HEAD	\
+				|| req->method_override == TFW_HTTP_METH_DELETE	\
+				|| req->method_override == TFW_HTTP_METH_TRACE)	\
+			{													\
+				TFW_PARSER_BLOCK(RGen_BodyInit);				\
+			}													\
+		}														\
+		else if (req->method == TFW_HTTP_METH_GET				\
+				 || req->method == TFW_HTTP_METH_HEAD			\
+				 || req->method == TFW_HTTP_METH_DELETE			\
+				 || req->method == TFW_HTTP_METH_TRACE)			\
+		{														\
+			TFW_PARSER_BLOCK(RGen_BodyInit);					\
+		}														\
+	}															\
 	if (msg->content_length) {					\
 		parser->to_read = msg->content_length;			\
 		__FSM_MOVE_nofixup(RGen_BodyStart);			\

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1155,13 +1155,21 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		 */							\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
-	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])	\
-		&& ((req->method == TFW_HTTP_METH_GET)	\
-		 || (req->method == TFW_HTTP_METH_HEAD)	\
-		 || (req->method == TFW_HTTP_METH_DELETE)	\
-		 || (req->method == TFW_HTTP_METH_TRACE)	\
-		   )) {	\
-		TFW_PARSER_BLOCK(RGen_BodyInit);			\
+	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])) {	\
+		/* Method override either honored or request message with method	\
+		 * override header dropped later in processing */	\
+		if (unlikely(req->method_override)) {	\
+			if (req->method_override == TFW_HTTP_METH_GET	\
+			 || req->method_override == TFW_HTTP_METH_HEAD	\
+			 || req->method_override == TFW_HTTP_METH_DELETE	\
+			 || req->method_override == TFW_HTTP_METH_TRACE)	\
+				TFW_PARSER_BLOCK(RGen_BodyInit);			\
+		} else if (req->method == TFW_HTTP_METH_GET	\
+			  || req->method == TFW_HTTP_METH_HEAD	\
+			  || req->method == TFW_HTTP_METH_DELETE	\
+			  || req->method == TFW_HTTP_METH_TRACE) {	\
+			TFW_PARSER_BLOCK(RGen_BodyInit);			\
+		}	\
 	}	\
 	if (msg->content_length) {					\
 		parser->to_read = msg->content_length;			\

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1155,6 +1155,10 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		 */							\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
+	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])	\
+		&& (req->method == TFW_HTTP_METH_GET)) {	\
+		TFW_PARSER_BLOCK(RGen_BodyInit);			\
+	}	\
 	if (msg->content_length) {					\
 		parser->to_read = msg->content_length;			\
 		__FSM_MOVE_nofixup(RGen_BodyStart);			\
@@ -8557,6 +8561,8 @@ tfw_h2_parse_req(void *req_data, unsigned char *data, size_t len,
 
 	if (likely(type != HTTP2_DATA))
 		r = tfw_hpack_decode(&ctx->hpack, data, len, req, parsed);
+	else if (unlikely(req->method == TFW_HTTP_METH_GET))
+		r = T_DROP;
 	else
 		r = tfw_h2_parse_body(data, len, req, parsed);
 

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1155,7 +1155,12 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		 */							\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
-	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])) {	\
+	/* According to RFC 7231 4.3.* a payload within GET, HEAD, DELETE,	\
+	 * TRACE and CONNECT requests has no defined semantics and	\
+	 * implementations can reject it. We do this respecting overrides.	\
+	*/	\
+	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])	\
+	 || !TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE])) {	\
 		/* Method override either honored or request message with method	\
 		 * override header dropped later in processing */	\
 		if (unlikely(req->method_override)) {	\

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -8489,7 +8489,7 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
 		__FSM_H2_DROP(Req_MethodUnknown);
 	}
 
-	/* Improbable states of shceme value processing. */
+	/* Improbable states of scheme value processing. */
 
 	__FSM_H2_SCHEME_STATE_MOVE(Req_Scheme_1CharStep, 'h', Req_SchemeH);
 	__FSM_H2_SCHEME_STATE_MOVE(Req_SchemeH, 't', Req_SchemeHt);

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1155,31 +1155,34 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		 */							\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
-	/* According to RFC 7231 4.3.* a payload within GET, HEAD, DELETE,	\
-	 * TRACE and CONNECT requests has no defined semantics and			\
-	 * implementations can reject it. We do this respecting overrides.	\
-	*/	\
+	/* According to RFC 7231 4.3.* a payload within GET, HEAD,	\
+	 * DELETE, TRACE and CONNECT requests has no defined semantics	\
+	 * and implementations can reject it. We do this respecting	\
+	 * overrides.							\
+	 */								\
 	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])		\
-		|| !TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE])) {	\
-		/* Method override either honored or request message with method	\
-		 * override header dropped later in processing */					\
-		if (unlikely(req->method_override)) {					\
+	    || !TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE]))		\
+	{								\
+		/* Method override either honored or request message	\
+		 * with method override header dropped later in		\
+		 * processing */					\
+		if (unlikely(req->method_override)) {			\
 			if (req->method_override == TFW_HTTP_METH_GET		\
-				|| req->method_override == TFW_HTTP_METH_HEAD	\
-				|| req->method_override == TFW_HTTP_METH_DELETE	\
-				|| req->method_override == TFW_HTTP_METH_TRACE)	\
-			{													\
-				TFW_PARSER_BLOCK(RGen_BodyInit);				\
-			}													\
-		}														\
-		else if (req->method == TFW_HTTP_METH_GET				\
-				 || req->method == TFW_HTTP_METH_HEAD			\
-				 || req->method == TFW_HTTP_METH_DELETE			\
-				 || req->method == TFW_HTTP_METH_TRACE)			\
-		{														\
-			TFW_PARSER_BLOCK(RGen_BodyInit);					\
-		}														\
-	}															\
+			    || req->method_override == TFW_HTTP_METH_HEAD	\
+			    || req->method_override == TFW_HTTP_METH_DELETE	\
+			    || req->method_override == TFW_HTTP_METH_TRACE)	\
+			{							\
+				TFW_PARSER_BLOCK(RGen_BodyInit);		\
+			}						\
+		}							\
+		else if (req->method == TFW_HTTP_METH_GET		\
+			 || req->method == TFW_HTTP_METH_HEAD		\
+			 || req->method == TFW_HTTP_METH_DELETE		\
+			 || req->method == TFW_HTTP_METH_TRACE)		\
+		{							\
+			TFW_PARSER_BLOCK(RGen_BodyInit);		\
+		}							\
+	}								\
 	if (msg->content_length) {					\
 		parser->to_read = msg->content_length;			\
 		__FSM_MOVE_nofixup(RGen_BodyStart);			\
@@ -8582,11 +8585,13 @@ tfw_h2_parse_req(void *req_data, unsigned char *data, size_t len,
 
 	if (likely(type != HTTP2_DATA))
 		r = tfw_hpack_decode(&ctx->hpack, data, len, req, parsed);
-	else if (unlikely(req->method == TFW_HTTP_METH_GET)
-		  || unlikely(req->method == TFW_HTTP_METH_HEAD)
-		  || unlikely(req->method == TFW_HTTP_METH_DELETE)
-		  || unlikely(req->method == TFW_HTTP_METH_TRACE))
+	else if (unlikely(req->method == TFW_HTTP_METH_GET
+		 || req->method == TFW_HTTP_METH_HEAD
+		 || req->method == TFW_HTTP_METH_DELETE
+		 || req->method == TFW_HTTP_METH_TRACE))
+	{
 		r = T_DROP;
+	}
 	else
 		r = tfw_h2_parse_body(data, len, req, parsed);
 

--- a/fw/http_parser.c
+++ b/fw/http_parser.c
@@ -1156,7 +1156,11 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}								\
 	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])	\
-		&& (req->method == TFW_HTTP_METH_GET)) {	\
+		&& ((req->method == TFW_HTTP_METH_GET)	\
+		 || (req->method == TFW_HTTP_METH_HEAD)	\
+		 || (req->method == TFW_HTTP_METH_DELETE)	\
+		 || (req->method == TFW_HTTP_METH_TRACE)	\
+		   )) {	\
 		TFW_PARSER_BLOCK(RGen_BodyInit);			\
 	}	\
 	if (msg->content_length) {					\
@@ -8561,7 +8565,10 @@ tfw_h2_parse_req(void *req_data, unsigned char *data, size_t len,
 
 	if (likely(type != HTTP2_DATA))
 		r = tfw_hpack_decode(&ctx->hpack, data, len, req, parsed);
-	else if (unlikely(req->method == TFW_HTTP_METH_GET))
+	else if (unlikely(req->method == TFW_HTTP_METH_GET)
+		  || unlikely(req->method == TFW_HTTP_METH_HEAD)
+		  || unlikely(req->method == TFW_HTTP_METH_DELETE)
+		  || unlikely(req->method == TFW_HTTP_METH_TRACE))
 		r = T_DROP;
 	else
 		r = tfw_h2_parse_body(data, len, req, parsed);

--- a/fw/t/fuzzer.h
+++ b/fw/t/fuzzer.h
@@ -30,7 +30,8 @@ enum {
 
 enum {
 	FUZZ_REQ,
-	FUZZ_RESP
+	FUZZ_RESP,
+	FUZZ_REQ_H2
 };
 
 typedef enum {

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -47,7 +47,7 @@
 #include "msg.c"
 #include "http_msg.c"
 
-static const int CHUNK_SIZES[] = { 1, 2, 3, 4, 8, 16, 32, 64, 128, 
+static const int CHUNK_SIZES[] = { 1, 2, 3, 4, 8, 16, 32, 64, 128,
                                    256, 1500, 9216, 1024*1024
                                   /* to fit a message of 'any' size */
                                  };
@@ -1482,7 +1482,12 @@ TEST(http_parser, content_length)
 	/* Content-Length is mandatory for responses. */
 	EXPECT_BLOCK_RESP("HTTP/1.1 200 OK\r\n\r\n");
 
-	FOR_REQ_SIMPLE("Content-Length: 0")
+	/* Content-Length must not be present in GET requests */
+	EXPECT_BLOCK_REQ_SIMPLE("Content-Length: 0");
+
+	FOR_REQ("POST / HTTP/1.1\r\n"
+		"Content-Length: 0\r\n"
+		"\r\n")
 	{
 		EXPECT_TRUE(req->content_length == 0);
 	}
@@ -1494,7 +1499,7 @@ TEST(http_parser, content_length)
 		EXPECT_TRUE(resp->content_length == 0);
 	}
 
-	FOR_REQ("GET / HTTP/1.1\r\n"
+	FOR_REQ("POST / HTTP/1.1\r\n"
 		"Content-Length: 5\r\n"
 		"\r\n"
 		"dummy")
@@ -1544,7 +1549,7 @@ TEST(http_parser, content_length)
 	EXPECT_BLOCK_DIGITS("HTTP/1.0 200 OK\r\nContent-Length: ",
 			    "\r\n\r\ndummy", EXPECT_BLOCK_RESP);
 
-	EXPECT_BLOCK_REQ("GET / HTTP/1.1\r\n"
+	EXPECT_BLOCK_REQ("POST / HTTP/1.1\r\n"
 			 "Content-Length: 10, 10\r\n"
 			 "\r\n"
 			 "0123456789");
@@ -1553,7 +1558,7 @@ TEST(http_parser, content_length)
 			  "\r\n"
 			  "0123456789");
 
-	EXPECT_BLOCK_REQ("GET / HTTP/1.1\r\n"
+	EXPECT_BLOCK_REQ("POST / HTTP/1.1\r\n"
 			 "Content-Length: 10 10\r\n"
 			 "\r\n"
 			 "0123456789");
@@ -1562,7 +1567,8 @@ TEST(http_parser, content_length)
 			  "\r\n"
 			  "0123456789");
 
-	EXPECT_BLOCK_REQ_SIMPLE("Content-Length: 0\r\n"
+	EXPECT_BLOCK_REQ("POST / HTTP/1.1"
+				"Content-Length: 0\r\n"
 		  		"Content-Length: 0");
 	EXPECT_BLOCK_RESP("HTTP/1.0 200 OK\r\n"
 			  "Content-Length: 0\r\n"

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -116,8 +116,10 @@ tfw_h2_pack_hdr_frame(const char *str, char buf[], unsigned int buf_len)
 		strsep(hdrp, ":");
 		hdr_val = hdrp ? strim(*hdrp) : "";
 		*(buf + H2_HDR_HDR_SZ + hdrs_len++) = LIT_HDR_FLD_WO_IND;
-		hdrs_len += tfw_h2_encode_str(hdr_one, buf + H2_HDR_HDR_SZ + hdrs_len);
-		hdrs_len += tfw_h2_encode_str(hdr_val, buf + H2_HDR_HDR_SZ + hdrs_len);
+		hdrs_len += tfw_h2_encode_str(hdr_one,
+					      buf + H2_HDR_HDR_SZ + hdrs_len);
+		hdrs_len += tfw_h2_encode_str(hdr_val,
+					      buf + H2_HDR_HDR_SZ + hdrs_len);
 	} while (*strp);
 
 	frame_hdr.length = hdrs_len;
@@ -330,49 +332,55 @@ validate_data_fully_parsed(int type)
 
 #define TRY_PARSE_EXPECT_PASS(str, type)			\
 ({ 								\
-	int _err = type == FUZZ_REQ_H2 ?	\
-		do_h2_parse(h2_buf + H2_HDR_HDR_SZ, h2_len, type) :	\
-		do_split_and_parse(str, type);		\
+	int _err = type == FUZZ_REQ_H2 ?			\
+		do_h2_parse(h2_buf + H2_HDR_HDR_SZ,		\
+			    h2_len, type) :			\
+		do_split_and_parse(str, type);			\
 	if (_err == TFW_BLOCK || _err == TFW_POSTPONE		\
 	    || !validate_data_fully_parsed(type))		\
 		TEST_FAIL("can't parse %s (code=%d):\n%s",	\
-			  (type == FUZZ_REQ || type == FUZZ_REQ_H2 ? "request" :	\
-				              "response"),	\
+			  (type == FUZZ_REQ			\
+			   || type == FUZZ_REQ_H2		\
+			   ? "request" : "response"),		\
 			  _err, (str)); 			\
 	__fpu_schedule();					\
-	type == FUZZ_REQ_H2 ? false:	\
-		_err == TFW_PASS;					\
+	type == FUZZ_REQ_H2 ? false:				\
+		_err == TFW_PASS;				\
 })
 
 #define TRY_PARSE_EXPECT_BLOCK(str, type)			\
 ({								\
-	int _err = type == FUZZ_REQ_H2 ?	\
-		do_h2_parse(h2_buf + H2_HDR_HDR_SZ, h2_len, type) :	\
-		do_split_and_parse(str, type);		\
+	int _err = type == FUZZ_REQ_H2 ?			\
+		do_h2_parse(h2_buf + H2_HDR_HDR_SZ,		\
+			    h2_len, type) :			\
+		do_split_and_parse(str, type);			\
 	if (_err == TFW_PASS)					\
 		TEST_FAIL("%s is not blocked as expected:\n%s",	\
-			       (type == FUZZ_REQ || type == FUZZ_REQ_H2 ? "request" :	\
-						   "response"),	\
+			  (type == FUZZ_REQ			\
+			   || type == FUZZ_REQ_H2		\
+			   ? "request" : "response"),		\
 			       (str));				\
 	__fpu_schedule();					\
-	type == FUZZ_REQ_H2 ? false :	\
+	type == FUZZ_REQ_H2 ? false :				\
 		_err == TFW_BLOCK || _err == TFW_POSTPONE;	\
 })
 
-#define __FOR_REQ(str, sz_diff, type)					\
+#define __FOR_REQ(str, sz_diff, type)				\
 	TEST_LOG("=== request: [%s]\n", str);			\
-	type == FUZZ_REQ_H2 ? test_case_parse_prepare_h2(str, sz_diff) :		\
-		test_case_parse_prepare_http(str, sz_diff);			\
+	type == FUZZ_REQ_H2 ?					\
+		test_case_parse_prepare_h2(str, sz_diff) :	\
+		test_case_parse_prepare_http(str, sz_diff);	\
 	while (TRY_PARSE_EXPECT_PASS(str, type))
 
 #define FOR_REQ(str)	__FOR_REQ(str, 0, FUZZ_REQ)
 #define FOR_REQ_H2(str)	__FOR_REQ(str, 0, FUZZ_REQ_H2)
 
-#define __EXPECT_BLOCK_REQ(str, type)					\
+#define __EXPECT_BLOCK_REQ(str, type)				\
 do {								\
 	TEST_LOG("=== request: [%s]\n", str);			\
-	type == FUZZ_REQ_H2 ? test_case_parse_prepare_h2(str, 0) :		\
-		test_case_parse_prepare_http(str, 0);			\
+	type == FUZZ_REQ_H2 ?					\
+		test_case_parse_prepare_h2(str, 0) :		\
+		test_case_parse_prepare_http(str, 0);		\
 	while (TRY_PARSE_EXPECT_BLOCK(str, type));		\
 } while (0)
 
@@ -381,7 +389,7 @@ do {								\
 
 #define __FOR_RESP(str, sz_diff)				\
 	TEST_LOG("=== response: [%s]\n", str);			\
-	test_case_parse_prepare_http(str, sz_diff);			\
+	test_case_parse_prepare_http(str, sz_diff);		\
 	while (TRY_PARSE_EXPECT_PASS(str, FUZZ_RESP))
 
 #define FOR_RESP(str)	__FOR_RESP(str, 0)
@@ -806,31 +814,31 @@ TEST(http_parser, mangled_messages)
 TEST(http_parser, alphabets)
 {
 	FOR_REQ("PUT / HTTP/1.1\r\n"
-			"Host: test\r\n"
-			/* We don't match open and closing quotes. */
-			"Content-Type: Text/HTML;Charset=utf-8\"\t  \n"
-			"Pragma: no-cache, fooo \r\n"
-			"\r\n");
+		"Host: test\r\n"
+		/* We don't match open and closing quotes. */
+		"Content-Type: Text/HTML;Charset=utf-8\"\t  \n"
+		"Pragma: no-cache, fooo \r\n"
+		"\r\n");
 
 	/* Trailing SP in request. */
 	FOR_REQ("PUT / HTTP/1.1\r\n"
-			"Host: localhost\t  \r\n"
-			"User-Agent: Wget/1.13.4 (linux-gnu)\t  \r\n"
-			"Accept: */*\t \r\n"
-			"Connection: Keep-Alive \t \r\n"
-			"X-Custom-Hdr: custom header values \t  \r\n"
-			"X-Forwarded-For: 127.0.0.1, example.com    \t \r\n"
-			"Content-Type: text/html; charset=iso-8859-1  \t \r\n"
-			"Cache-Control: "
-			"max-age=0, private, min-fresh=42 \t \r\n"
-			"Transfer-Encoding: "
-			"compress, deflate, gzip, chunked\t  \r\n"
-			"Cookie: session=42; theme=dark  \t \r\n"
-			"\r\n"
-			"3\r\n"
-			"123\r\n"
-			"0\r\n"
-			"\r\n");
+		"Host: localhost\t  \r\n"
+		"User-Agent: Wget/1.13.4 (linux-gnu)\t  \r\n"
+		"Accept: */*\t \r\n"
+		"Connection: Keep-Alive \t \r\n"
+		"X-Custom-Hdr: custom header values \t  \r\n"
+		"X-Forwarded-For: 127.0.0.1, example.com    \t \r\n"
+		"Content-Type: text/html; charset=iso-8859-1  \t \r\n"
+		"Cache-Control: "
+		"max-age=0, private, min-fresh=42 \t \r\n"
+		"Transfer-Encoding: "
+		"compress, deflate, gzip, chunked\t  \r\n"
+		"Cookie: session=42; theme=dark  \t \r\n"
+		"\r\n"
+		"3\r\n"
+		"123\r\n"
+		"0\r\n"
+		"\r\n");
 
 	/* Trailing SP in response. */
 	FOR_RESP("HTTP/1.1 200 OK\r\n"
@@ -855,11 +863,11 @@ TEST(http_parser, alphabets)
 TEST(http_parser, casesense)
 {
 	FOR_REQ("PUT / HTTP/1.1\r\n"
-			"hOST: test\r\n"
-			"cAchE-CoNtRoL: no-cache\n"
-			"x-fORWarDED-For: 1.1.1.1\r\n"
-			"conTent-typE: chunked\r\n"
-			"\r\n");
+		"hOST: test\r\n"
+		"cAchE-CoNtRoL: no-cache\n"
+		"x-fORWarDED-For: 1.1.1.1\r\n"
+		"conTent-typE: chunked\r\n"
+		"\r\n");
 
 	FOR_RESP("HTTP/1.1 200 OK\r\n"
 		"aGE: 10\r\n"
@@ -1609,23 +1617,23 @@ TEST(http_parser, content_type_in_bodyless_requests)
 	EXPECT_BLOCK_REQ_SIMPLE("Content-Type: text/plain");
 
 	EXPECT_BLOCK_REQ("HEAD / HTTP/1.1\r\n"
-			  "Content-Type: text/html; charset=utf-8\r\n"
-		 	  "Content-Length: 0\r\n"
-		 	  "\r\n");
+			 "Content-Type: text/html; charset=utf-8\r\n"
+		 	 "Content-Length: 0\r\n"
+		 	 "\r\n");
 
 	EXPECT_BLOCK_REQ("DELETE / HTTP/1.1\r\n"
-			  "Content-Length: 5\r\n"
-			  "Content-Type: text/html\r\n"
-			  "\r\n"
-			  "dummy");
+			 "Content-Length: 5\r\n"
+			 "Content-Type: text/html\r\n"
+			 "\r\n"
+			 "dummy");
 
 	EXPECT_BLOCK_REQ("TRACE / HTTP/1.1\r\n"
-			  "Content-Type: application/octet-stream\r\n"
-			  "\r\n");
+			 "Content-Type: application/octet-stream\r\n"
+			 "\r\n");
 
 	FOR_REQ("OPTIONS / HTTP/1.1\r\n"
-			"Content-Type: text/plain\r\n"
-			"\r\n")
+		"Content-Type: text/plain\r\n"
+		"\r\n")
 	{
 		EXPECT_TFWSTR_EQ(&req->h_tbl->tbl[TFW_HTTP_HDR_CONTENT_TYPE],
 				 "Content-Type: text/plain");
@@ -1633,15 +1641,15 @@ TEST(http_parser, content_type_in_bodyless_requests)
 
 	/* Without content-length will not be blocked */
 	FOR_REQ_H2(":method: GET\n"
-			   ":scheme: https\n"
-			   ":path: /");
+		   ":scheme: https\n"
+		   ":path: /");
 
 	/* But with content-length will be block for http2 too */
 	EXPECT_BLOCK_REQ_H2(":authority: debian\n"
-						":method: GET\n"
-						":scheme: http\n"
-						":path: /\n"
-						"content-length: 0");
+			    ":method: GET\n"
+			    ":scheme: http\n"
+			    ":path: /\n"
+			    "content-length: 0");
 }
 
 TEST(http_parser, content_length)

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -69,13 +69,65 @@ static size_t hm_exp_len = 0;
 				TOKEN_ALPHABET OBS_TEXT
 #define ETAG_ALPHABET		OTHER_DELIMETERS TOKEN_ALPHABET OBS_TEXT
 
+static
+unsigned int
+tfw_h2_encode_str(const char *str, char *buf)
+{
+	TfwHPackInt hpint;
+	unsigned int len = strlen(str);
+	write_int(len, 0x7f, 0, &hpint);
+	memcpy_fast(buf, hpint.buf, hpint.sz);
+	memcpy_fast(buf + hpint.sz, str, len);
+	return hpint.sz + len;
+}
+
+#define H2_HDR_HDR_SZ 9
+static
+unsigned int
+tfw_h2_pack_hdr_frame(const char *str, char buf[], unsigned int buf_len)
+{
+	unsigned long hdrs_len = 0;
+	TfwFrameHdr frame_hdr = {.flags = HTTP2_F_END_HEADERS | HTTP2_F_END_STREAM,
+							 .type = HTTP2_HEADERS,
+							 .stream_id = 1};
+	char **strp, *hdr_one, *hdr, **hdrp, *hdr_val;
+	char str_buf[256] = {0};
+	char *_str = str_buf;
+
+	BUG_ON(strlen(str) + 1 > ARRAY_SIZE(str_buf));
+
+	memcpy_fast(_str, str, strlen(str));
+
+	strp = &_str;
+	do {
+		hdr_one = strsep(strp, "\n");
+		hdr = hdr_one;
+		if (*hdr == ':') {
+			++hdr;
+			if (!*hdr)
+				continue;
+		}
+		hdrp = &hdr;
+		strsep(hdrp, ":");
+		hdr_val = hdrp ? strim(*hdrp) : "";
+		hdrs_len += tfw_h2_encode_str(hdr_one, buf + H2_HDR_HDR_SZ + hdrs_len);
+		hdrs_len += tfw_h2_encode_str(hdr_val, buf + H2_HDR_HDR_SZ + hdrs_len);
+	} while (*strp);
+
+	frame_hdr.length = hdrs_len;
+	tfw_h2_pack_frame_header(buf, &frame_hdr);
+
+	BUG_ON(frame_hdr.length + H2_HDR_HDR_SZ > buf_len);
+	return frame_hdr.length + H2_HDR_HDR_SZ;
+}
+
 static int
 split_and_parse_n(unsigned char *str, int type, size_t len, size_t chunk_size)
 {
 	size_t pos = 0;
 	unsigned int parsed;
 	int r = 0;
-	TfwHttpMsg *hm = (type == FUZZ_REQ)
+	TfwHttpMsg *hm = (type == FUZZ_REQ || type == FUZZ_REQ_H2)
 			? (TfwHttpMsg *)req
 			: (TfwHttpMsg *)resp;
 
@@ -87,6 +139,8 @@ split_and_parse_n(unsigned char *str, int type, size_t len, size_t chunk_size)
 			  len, pos);
 		if (type == FUZZ_REQ)
 			r = tfw_http_parse_req(req, str + pos, chunk_size, &parsed);
+		else if (type == FUZZ_REQ_H2)
+			r = tfw_h2_parse_req(req, str + pos, chunk_size, &parsed);
 		else
 			r = tfw_http_parse_resp(resp, str + pos, chunk_size, &parsed);
 
@@ -130,7 +184,7 @@ test_case_parse_prepare(const char *str, size_t sz_diff)
  * The function is designed to be called in a loop, e.g.
  *   while(!do_split_and_parse(str, type)) { ... }
  *
- * type may be FUZZ_REQ or FUZZ_RESP.
+ * type may be FUZZ_REQ or FUZZ_REQ_H2 or FUZZ_RESP.
  *
  * On each iteration it splits the @str into fragments and pushes
  * them to the HTTP parser.
@@ -158,6 +212,7 @@ do_split_and_parse(unsigned char *str, int type)
 {
 	int r;
 	static size_t len;
+	static char h2_buf[1024];
 
 	BUG_ON(!str);
 
@@ -171,7 +226,7 @@ do_split_and_parse(unsigned char *str, int type)
 		 */
 		return 1;
 
-	if (type == FUZZ_REQ) {
+	if (type == FUZZ_REQ || type == FUZZ_REQ_H2) {
 		if (req)
 			test_req_free(req);
 
@@ -188,6 +243,10 @@ do_split_and_parse(unsigned char *str, int type)
 		BUG();
 	}
 
+	if (type == FUZZ_REQ_H2) {
+		len = tfw_h2_pack_hdr_frame(str, h2_buf, ARRAY_SIZE(h2_buf));
+		str = h2_buf;
+	}
 	r = split_and_parse_n(str, type, len, CHUNK_SIZES[chunk_size_index]);
 
 	if (CHUNK_SIZES[chunk_size_index] >= len)
@@ -225,7 +284,7 @@ validate_data_fully_parsed(int type)
 	if (_err == TFW_BLOCK || _err == TFW_POSTPONE		\
 	    || !validate_data_fully_parsed(type))		\
 		TEST_FAIL("can't parse %s (code=%d):\n%s",	\
-			  (type == FUZZ_REQ ? "request" :	\
+			  (type == FUZZ_REQ || type == FUZZ_REQ_H2 ? "request" :	\
 				              "response"),	\
 			  _err, (str)); 			\
 	__fpu_schedule();					\
@@ -237,26 +296,30 @@ validate_data_fully_parsed(int type)
 	int _err = do_split_and_parse(str, type);		\
 	if (_err == TFW_PASS)					\
 		TEST_FAIL("%s is not blocked as expected:\n%s",	\
-			       (type == FUZZ_REQ ? "request" :	\
+			       (type == FUZZ_REQ || type == FUZZ_REQ_H2 ? "request" :	\
 						   "response"),	\
 			       (str));				\
 	__fpu_schedule();					\
 	_err == TFW_BLOCK || _err == TFW_POSTPONE;		\
 })
 
-#define __FOR_REQ(str, sz_diff)					\
+#define __FOR_REQ(str, sz_diff, type)					\
 	TEST_LOG("=== request: [%s]\n", str);			\
 	test_case_parse_prepare(str, sz_diff);			\
-	while (TRY_PARSE_EXPECT_PASS(str, FUZZ_REQ))
+	while (TRY_PARSE_EXPECT_PASS(str, type))
 
-#define FOR_REQ(str)	__FOR_REQ(str, 0)
+#define FOR_REQ(str)	__FOR_REQ(str, 0, FUZZ_REQ)
+#define FOR_REQ_H2(str)	__FOR_REQ(str, 0, FUZZ_REQ_H2)
 
-#define EXPECT_BLOCK_REQ(str)					\
+#define __EXPECT_BLOCK_REQ(str, type)					\
 do {								\
 	TEST_LOG("=== request: [%s]\n", str);			\
 	test_case_parse_prepare(str, 0);			\
-	while (TRY_PARSE_EXPECT_BLOCK(str, FUZZ_REQ));		\
+	while (TRY_PARSE_EXPECT_BLOCK(str, type));		\
 } while (0)
+
+#define EXPECT_BLOCK_REQ(str)		__EXPECT_BLOCK_REQ(str, FUZZ_REQ)
+#define EXPECT_BLOCK_REQ_H2(str)	__EXPECT_BLOCK_REQ(str, FUZZ_REQ_H2)
 
 #define __FOR_RESP(str, sz_diff)				\
 	TEST_LOG("=== response: [%s]\n", str);			\
@@ -1518,6 +1581,17 @@ TEST(http_parser, content_type_in_bodyless_requests)
 
 TEST(http_parser, content_length)
 {
+	// EXPECT_BLOCK_REQ_H2(":authority: debian\n"
+	// 					":method: GET\n"
+	// 					":scheme: http\n"
+	// 					":path: /\n"
+	// 					"content-length: 0");
+
+	// FOR_REQ_H2(":authority: debian\n"
+	// 			":method: GET\n"
+	// 			":scheme: http\n"
+	// 			":path: /");
+
 	/* Content-Length is mandatory for responses. */
 	EXPECT_BLOCK_RESP("HTTP/1.1 200 OK\r\n\r\n");
 
@@ -1688,7 +1762,7 @@ TEST(http_parser, eol_crlf)
 		  "\n"
 		  "a=24\n"
 		  "\n",  /* the LF is ignored. */
-		  1)
+		  1, FUZZ_REQ)
 	{
 		TfwHttpHdrTbl *ht = req->h_tbl;
 
@@ -1704,7 +1778,7 @@ TEST(http_parser, eol_crlf)
 		  "\r\n"
 		  "b=24\r\n"
 		  "\r\n",  /* the CRLF is ignored. */
-		  2)
+		  2, FUZZ_REQ)
 	{
 		EXPECT_TRUE(req->crlf.len == 2);
 		EXPECT_TRUE(req->body.len == 6);

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -56,6 +56,11 @@ static unsigned int chunk_size_index = 0;
 
 static TfwHttpReq *req, *sample_req;
 static TfwHttpResp *resp;
+static TfwH2Conn conn;
+static TfwStream stream;
+static TfwStr h2_parsed_hdr;
+static char h2_buf[1024];
+static unsigned int h2_len = 0;
 static size_t hm_exp_len = 0;
 
 #define SAMPLE_REQ_STR	"GET / HTTP/1.1\r\nHost: example.com\r\n\r\n"
@@ -82,6 +87,7 @@ tfw_h2_encode_str(const char *str, char *buf)
 }
 
 #define H2_HDR_HDR_SZ 9
+#define LIT_HDR_FLD_WO_IND 0x00
 static
 unsigned int
 tfw_h2_pack_hdr_frame(const char *str, char buf[], unsigned int buf_len)
@@ -110,6 +116,7 @@ tfw_h2_pack_hdr_frame(const char *str, char buf[], unsigned int buf_len)
 		hdrp = &hdr;
 		strsep(hdrp, ":");
 		hdr_val = hdrp ? strim(*hdrp) : "";
+		*(buf + H2_HDR_HDR_SZ + hdrs_len++) = LIT_HDR_FLD_WO_IND;
 		hdrs_len += tfw_h2_encode_str(hdr_one, buf + H2_HDR_HDR_SZ + hdrs_len);
 		hdrs_len += tfw_h2_encode_str(hdr_val, buf + H2_HDR_HDR_SZ + hdrs_len);
 	} while (*strp);
@@ -127,10 +134,11 @@ split_and_parse_n(unsigned char *str, int type, size_t len, size_t chunk_size)
 	size_t pos = 0;
 	unsigned int parsed;
 	int r = 0;
-	TfwHttpMsg *hm = (type == FUZZ_REQ || type == FUZZ_REQ_H2)
+	TfwHttpMsg *hm = (type == FUZZ_REQ)
 			? (TfwHttpMsg *)req
 			: (TfwHttpMsg *)resp;
 
+	BUG_ON(type != FUZZ_REQ && type != FUZZ_RESP);
 	while (pos < len) {
 		if (chunk_size >= len - pos)
 			/* At the last chunk */
@@ -139,8 +147,6 @@ split_and_parse_n(unsigned char *str, int type, size_t len, size_t chunk_size)
 			  len, pos);
 		if (type == FUZZ_REQ)
 			r = tfw_http_parse_req(req, str + pos, chunk_size, &parsed);
-		else if (type == FUZZ_REQ_H2)
-			r = tfw_h2_parse_req(req, str + pos, chunk_size, &parsed);
 		else
 			r = tfw_http_parse_resp(resp, str + pos, chunk_size, &parsed);
 
@@ -174,10 +180,28 @@ set_sample_req(unsigned char *str)
 }
 
 static void
-test_case_parse_prepare(const char *str, size_t sz_diff)
+test_case_parse_prepare_http(const char *str, size_t sz_diff)
 {
 	chunk_size_index = 0;
 	hm_exp_len = strlen(str) - sz_diff;
+}
+
+static void
+test_case_parse_prepare_h2(const char *str, size_t sz_diff)
+{
+	h2_len = tfw_h2_pack_hdr_frame(str, h2_buf, ARRAY_SIZE(h2_buf));
+	h2_len -= H2_HDR_HDR_SZ;
+
+	tfw_h2_context_init(&conn.h2);
+	conn.h2.hdr.type = HTTP2_HEADERS;
+	h2_parsed_hdr = (TfwStr){
+		.data = h2_buf,
+		.len = H2_HDR_HDR_SZ,
+		.nchunks = 1
+	};
+	stream.state = HTTP2_STREAM_REM_HALF_CLOSED;
+
+	hm_exp_len = h2_len - sz_diff;
 }
 
 /**
@@ -212,7 +236,6 @@ do_split_and_parse(unsigned char *str, int type)
 {
 	int r;
 	static size_t len;
-	static char h2_buf[1024];
 
 	BUG_ON(!str);
 
@@ -226,7 +249,7 @@ do_split_and_parse(unsigned char *str, int type)
 		 */
 		return 1;
 
-	if (type == FUZZ_REQ || type == FUZZ_REQ_H2) {
+	if (type == FUZZ_REQ) {
 		if (req)
 			test_req_free(req);
 
@@ -243,10 +266,6 @@ do_split_and_parse(unsigned char *str, int type)
 		BUG();
 	}
 
-	if (type == FUZZ_REQ_H2) {
-		len = tfw_h2_pack_hdr_frame(str, h2_buf, ARRAY_SIZE(h2_buf));
-		str = h2_buf;
-	}
 	r = split_and_parse_n(str, type, len, CHUNK_SIZES[chunk_size_index]);
 
 	if (CHUNK_SIZES[chunk_size_index] >= len)
@@ -262,6 +281,38 @@ do_split_and_parse(unsigned char *str, int type)
 	return r;
 }
 
+static int
+do_h2_parse(unsigned char *str, unsigned int len, int type)
+{
+	int r;
+	unsigned int parsed;
+
+	BUG_ON(!str);
+
+	if (type == FUZZ_REQ_H2) {
+		if (req)
+			test_req_free(req);
+
+		req = test_req_alloc(len);
+	}
+	else {
+		BUG();
+	}
+
+	req->conn = (TfwConn*)&conn;
+	req->pit.parsed_hdr = &h2_parsed_hdr;
+	stream.msg = (TfwMsg*)req;
+	req->stream = &stream;
+
+	r = tfw_h2_parse_req(req, str, len, &parsed);
+
+	((TfwHttpMsg*)req)->msg.len += parsed;
+	if (r == T_POSTPONE)
+		r = tfw_h2_parse_req_finish(req);
+
+	return r;
+}
+
 /**
  * To validate message parsing we provide text string which describes
  * HTTP message from start to end. If there any unused bytes after
@@ -270,7 +321,7 @@ do_split_and_parse(unsigned char *str, int type)
 static int
 validate_data_fully_parsed(int type)
 {
-	TfwHttpMsg *hm = (type == FUZZ_REQ)
+	TfwHttpMsg *hm = (type == FUZZ_REQ || type == FUZZ_REQ_H2)
 			? (TfwHttpMsg *)req
 			: (TfwHttpMsg *)resp;
 
@@ -280,7 +331,9 @@ validate_data_fully_parsed(int type)
 
 #define TRY_PARSE_EXPECT_PASS(str, type)			\
 ({ 								\
-	int _err = do_split_and_parse(str, type);		\
+	int _err = type == FUZZ_REQ_H2 ?	\
+		do_h2_parse(h2_buf + H2_HDR_HDR_SZ, h2_len, type) :	\
+		do_split_and_parse(str, type);		\
 	if (_err == TFW_BLOCK || _err == TFW_POSTPONE		\
 	    || !validate_data_fully_parsed(type))		\
 		TEST_FAIL("can't parse %s (code=%d):\n%s",	\
@@ -288,24 +341,29 @@ validate_data_fully_parsed(int type)
 				              "response"),	\
 			  _err, (str)); 			\
 	__fpu_schedule();					\
-	_err == TFW_PASS;					\
+	type == FUZZ_REQ_H2 ? false:	\
+		_err == TFW_PASS;					\
 })
 
 #define TRY_PARSE_EXPECT_BLOCK(str, type)			\
 ({								\
-	int _err = do_split_and_parse(str, type);		\
+	int _err = type == FUZZ_REQ_H2 ?	\
+		do_h2_parse(h2_buf + H2_HDR_HDR_SZ, h2_len, type) :	\
+		do_split_and_parse(str, type);		\
 	if (_err == TFW_PASS)					\
 		TEST_FAIL("%s is not blocked as expected:\n%s",	\
 			       (type == FUZZ_REQ || type == FUZZ_REQ_H2 ? "request" :	\
 						   "response"),	\
 			       (str));				\
 	__fpu_schedule();					\
-	_err == TFW_BLOCK || _err == TFW_POSTPONE;		\
+	type == FUZZ_REQ_H2 ? false :	\
+		_err == TFW_BLOCK || _err == TFW_POSTPONE;	\
 })
 
 #define __FOR_REQ(str, sz_diff, type)					\
 	TEST_LOG("=== request: [%s]\n", str);			\
-	test_case_parse_prepare(str, sz_diff);			\
+	type == FUZZ_REQ_H2 ? test_case_parse_prepare_h2(str, sz_diff) :		\
+		test_case_parse_prepare_http(str, sz_diff);			\
 	while (TRY_PARSE_EXPECT_PASS(str, type))
 
 #define FOR_REQ(str)	__FOR_REQ(str, 0, FUZZ_REQ)
@@ -314,7 +372,8 @@ validate_data_fully_parsed(int type)
 #define __EXPECT_BLOCK_REQ(str, type)					\
 do {								\
 	TEST_LOG("=== request: [%s]\n", str);			\
-	test_case_parse_prepare(str, 0);			\
+	type == FUZZ_REQ_H2 ? test_case_parse_prepare_h2(str, 0) :		\
+		test_case_parse_prepare_http(str, 0);			\
 	while (TRY_PARSE_EXPECT_BLOCK(str, type));		\
 } while (0)
 
@@ -323,7 +382,7 @@ do {								\
 
 #define __FOR_RESP(str, sz_diff)				\
 	TEST_LOG("=== response: [%s]\n", str);			\
-	test_case_parse_prepare(str, sz_diff);			\
+	test_case_parse_prepare_http(str, sz_diff);			\
 	while (TRY_PARSE_EXPECT_PASS(str, FUZZ_RESP))
 
 #define FOR_RESP(str)	__FOR_RESP(str, 0)
@@ -331,7 +390,7 @@ do {								\
 #define EXPECT_BLOCK_RESP(str)					\
 do {								\
 	TEST_LOG("=== response: [%s]\n", str);			\
-	test_case_parse_prepare(str, 0);			\
+	test_case_parse_prepare_http(str, 0);			\
 	while (TRY_PARSE_EXPECT_BLOCK(str, FUZZ_RESP));		\
 } while (0)
 
@@ -1577,21 +1636,22 @@ TEST(http_parser, content_type_in_bodyless_requests)
 		EXPECT_TFWSTR_EQ(&req->h_tbl->tbl[TFW_HTTP_HDR_CONTENT_TYPE],
 				 "Content-Type: text/plain");
 	}
+
+	/* Without content-length will not be blocked */
+	FOR_REQ_H2(":method: GET\n"
+			   ":scheme: https\n"
+			   ":path: /");
+
+	/* But wiht content-length will be block for http2 too */
+	EXPECT_BLOCK_REQ_H2(":authority: debian\n"
+						":method: GET\n"
+						":scheme: http\n"
+						":path: /\n"
+						"content-length: 0");
 }
 
 TEST(http_parser, content_length)
 {
-	// EXPECT_BLOCK_REQ_H2(":authority: debian\n"
-	// 					":method: GET\n"
-	// 					":scheme: http\n"
-	// 					":path: /\n"
-	// 					"content-length: 0");
-
-	// FOR_REQ_H2(":authority: debian\n"
-	// 			":method: GET\n"
-	// 			":scheme: http\n"
-	// 			":path: /");
-
 	/* Content-Length is mandatory for responses. */
 	EXPECT_BLOCK_RESP("HTTP/1.1 200 OK\r\n\r\n");
 
@@ -1724,10 +1784,10 @@ TEST(http_parser, content_length)
 		 "Content-Length: 0\r\n"
 		 "\r\n");
 
-	/*
-	 * A server MAY send a Content-Length header field in a 304
-	 * (Not Modified) response to a conditional GET request.
-	 */
+		/*
+		* A server MAY send a Content-Length header field in a 304
+		* (Not Modified) response to a conditional GET request.
+		*/
 #define NOT_PARSED "dummy_body"
 #define RESP	"HTTP/1.1 304 Not Modified\r\n"		\
 		"Content-Length: 5\r\n"			\
@@ -3331,11 +3391,11 @@ TEST(http_parser, fuzzer)
 				       FUZZ_REQ);
 			switch (ret) {
 			case FUZZ_VALID:
-				test_case_parse_prepare(str, 0);
+				test_case_parse_prepare_http(str, 0);
 				TRY_PARSE_EXPECT_PASS(str, FUZZ_REQ);
 				break;
 			case FUZZ_INVALID:
-				test_case_parse_prepare(str, 0);
+				test_case_parse_prepare_http(str, 0);
 				TRY_PARSE_EXPECT_BLOCK(str, FUZZ_REQ);
 				break;
 			case FUZZ_END:
@@ -3357,11 +3417,11 @@ resp:
 				       FUZZ_RESP);
 			switch (ret) {
 			case FUZZ_VALID:
-				test_case_parse_prepare(str, 0);
+				test_case_parse_prepare_http(str, 0);
 				TRY_PARSE_EXPECT_PASS(str, FUZZ_RESP);
 				break;
 			case FUZZ_INVALID:
-				test_case_parse_prepare(str, 0);
+				test_case_parse_prepare_http(str, 0);
 				TRY_PARSE_EXPECT_BLOCK(str, FUZZ_RESP);
 				break;
 			case FUZZ_END:
@@ -4097,7 +4157,7 @@ TEST(http_parser, perf)
 
 #define REQ_PERF(str)							\
 do {									\
-	test_case_parse_prepare(str, 0);				\
+	test_case_parse_prepare_http(str, 0);				\
 	if (req)							\
 		test_req_free(req);					\
 	req = test_req_alloc(sizeof(str) - 1);				\
@@ -4106,7 +4166,7 @@ do {									\
 
 #define RESP_PERF(str)							\
 do {									\
-	test_case_parse_prepare(str, 0);				\
+	test_case_parse_prepare_http(str, 0);				\
 	if (resp)							\
 		test_resp_free(resp);					\
 	resp = test_resp_alloc(sizeof(str) - 1);			\

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -109,8 +109,7 @@ tfw_h2_pack_hdr_frame(const char *str, char buf[], unsigned int buf_len)
 		hdr_one = strsep(strp, "\n");
 		hdr = hdr_one;
 		if (*hdr == ':') {
-			++hdr;
-			if (!*hdr)
+			if (!*(++hdr))
 				continue;
 		}
 		hdrp = &hdr;
@@ -1642,7 +1641,7 @@ TEST(http_parser, content_type_in_bodyless_requests)
 			   ":scheme: https\n"
 			   ":path: /");
 
-	/* But wiht content-length will be block for http2 too */
+	/* But with content-length will be block for http2 too */
 	EXPECT_BLOCK_REQ_H2(":authority: debian\n"
 						":method: GET\n"
 						":scheme: http\n"
@@ -1784,10 +1783,10 @@ TEST(http_parser, content_length)
 		 "Content-Length: 0\r\n"
 		 "\r\n");
 
-		/*
-		* A server MAY send a Content-Length header field in a 304
-		* (Not Modified) response to a conditional GET request.
-		*/
+	/*
+	 * A server MAY send a Content-Length header field in a 304
+	 * (Not Modified) response to a conditional GET request.
+	 */
 #define NOT_PARSED "dummy_body"
 #define RESP	"HTTP/1.1 304 Not Modified\r\n"		\
 		"Content-Length: 5\r\n"			\

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -684,28 +684,32 @@ TEST(http_parser, mangled_messages)
  */
 TEST(http_parser, alphabets)
 {
-	FOR_REQ_SIMPLE("Host: test\r\n"
-		       /* We don't match open and closing quotes. */
-		       "Content-Type: Text/HTML;Charset=utf-8\"\t  \n"
-		       "Pragma: no-cache, fooo ");
+	FOR_REQ("PUT / HTTP/1.1\r\n"
+			"Host: test\r\n"
+			/* We don't match open and closing quotes. */
+			"Content-Type: Text/HTML;Charset=utf-8\"\t  \n"
+			"Pragma: no-cache, fooo \r\n"
+			"\r\n");
 
 	/* Trailing SP in request. */
-	FOR_REQ_SIMPLE("Host: localhost\t  \r\n"
-		       "User-Agent: Wget/1.13.4 (linux-gnu)\t  \r\n"
-		       "Accept: */*\t \r\n"
-		       "Connection: Keep-Alive \t \r\n"
-		       "X-Custom-Hdr: custom header values \t  \r\n"
-		       "X-Forwarded-For: 127.0.0.1, example.com    \t \r\n"
-		       "Content-Type: text/html; charset=iso-8859-1  \t \r\n"
-		       "Cache-Control: "
-		       "max-age=0, private, min-fresh=42 \t \r\n"
-		       "Transfer-Encoding: "
-		       "compress, deflate, gzip, chunked\t  \r\n"
-		       "Cookie: session=42; theme=dark  \t \r\n"
-		       "\r\n"
-		       "3\r\n"
-		       "123\r\n"
-		       "0");
+	FOR_REQ("PUT / HTTP/1.1\r\n"
+			"Host: localhost\t  \r\n"
+			"User-Agent: Wget/1.13.4 (linux-gnu)\t  \r\n"
+			"Accept: */*\t \r\n"
+			"Connection: Keep-Alive \t \r\n"
+			"X-Custom-Hdr: custom header values \t  \r\n"
+			"X-Forwarded-For: 127.0.0.1, example.com    \t \r\n"
+			"Content-Type: text/html; charset=iso-8859-1  \t \r\n"
+			"Cache-Control: "
+			"max-age=0, private, min-fresh=42 \t \r\n"
+			"Transfer-Encoding: "
+			"compress, deflate, gzip, chunked\t  \r\n"
+			"Cookie: session=42; theme=dark  \t \r\n"
+			"\r\n"
+			"3\r\n"
+			"123\r\n"
+			"0\r\n"
+			"\r\n");
 
 	/* Trailing SP in response. */
 	FOR_RESP("HTTP/1.1 200 OK\r\n"
@@ -729,10 +733,12 @@ TEST(http_parser, alphabets)
  */
 TEST(http_parser, casesense)
 {
-	FOR_REQ_SIMPLE("hOST: test\r\n"
-		       "cAchE-CoNtRoL: no-cache\n"
-		       "x-fORWarDED-For: 1.1.1.1\r\n"
-		       "conTent-typE: chunked");
+	FOR_REQ("PUT / HTTP/1.1\r\n"
+			"hOST: test\r\n"
+			"cAchE-CoNtRoL: no-cache\n"
+			"x-fORWarDED-For: 1.1.1.1\r\n"
+			"conTent-typE: chunked\r\n"
+			"\r\n");
 
 	FOR_RESP("HTTP/1.1 200 OK\r\n"
 		"aGE: 10\r\n"
@@ -1474,6 +1480,39 @@ TEST(http_parser, parses_connection_value)
 		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_CLOSE, req->flags));
 		EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
 		EXPECT_TRUE(test_bit(TFW_HTTP_B_CONN_EXTRA, req->flags));
+	}
+}
+
+TEST(http_parser, content_type_in_bodyless_requests)
+{
+	EXPECT_BLOCK_REQ_SIMPLE("Content-Type: text/plain");
+
+	EXPECT_BLOCK_REQ("HEAD / HTTP/1.1\r\n"
+			  "Content-Type: text/html; charset=utf-8\r\n"
+		 	  "Content-Length: 0\r\n"
+		 	  "\r\n");
+
+	EXPECT_BLOCK_REQ("DELETE / HTTP/1.1\r\n"
+			  "Content-Length: 5\r\n"
+			  "Content-Type: text/html\r\n"
+			  "\r\n"
+			  "dummy");
+
+	EXPECT_BLOCK_REQ("TRACE / HTTP/1.1\r\n"
+			  "Content-Type: application/octet-stream\r\n"
+			  "\r\n");
+
+	/* Uncomment when CONNECT method will be added to Tempesta FW */
+	// EXPECT_BLOCK_REQ("CONNECT / HTTP/1.1\r\n"
+	// 		  "Content-Type: application/octet-stream\r\n"
+	// 		  "\r\n");
+
+	FOR_REQ("OPTIONS / HTTP/1.1\r\n"
+			"Content-Type: text/plain\r\n"
+			"\r\n")
+	{
+		EXPECT_TFWSTR_EQ(&req->h_tbl->tbl[TFW_HTTP_HDR_CONTENT_TYPE],
+				 "Content-Type: text/plain");
 	}
 }
 
@@ -4122,6 +4161,7 @@ TEST_SUITE(http_parser)
 	TEST_RUN(http_parser, pragma);
 	TEST_RUN(http_parser, suspicious_x_forwarded_for);
 	TEST_RUN(http_parser, parses_connection_value);
+	TEST_RUN(http_parser, content_type_in_bodyless_requests);
 	TEST_RUN(http_parser, content_length);
 	TEST_RUN(http_parser, eol_crlf);
 	TEST_RUN(http_parser, ows);

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -1623,11 +1623,6 @@ TEST(http_parser, content_type_in_bodyless_requests)
 			  "Content-Type: application/octet-stream\r\n"
 			  "\r\n");
 
-	/* Uncomment when CONNECT method will be added to Tempesta FW */
-	// EXPECT_BLOCK_REQ("CONNECT / HTTP/1.1\r\n"
-	// 		  "Content-Type: application/octet-stream\r\n"
-	// 		  "\r\n");
-
 	FOR_REQ("OPTIONS / HTTP/1.1\r\n"
 			"Content-Type: text/plain\r\n"
 			"\r\n")
@@ -3064,7 +3059,7 @@ TEST(http_parser, req_hop_by_hop)
 	TfwStr *field;
 	long id;
 #define REQ_HBH_START							\
-	"GET /foo HTTP/1.1\r\n"						\
+	"POST /foo HTTP/1.1\r\n"						\
 	"User-Agent: Wget/1.13.4 (linux-gnu)\r\n"			\
 	"Accept: */*\r\n"						\
 	"Host: localhost\r\n"						\
@@ -3082,9 +3077,12 @@ TEST(http_parser, req_hop_by_hop)
 	"Dummy5: 5\r\n"							\
 	"Foo: is hop-by-hop header\r\n"					\
 	"Dummy6: 6\r\n"							\
-	"Buzz: is hop-by-hop header\r\n"				\
+	"Content-Length: 0\r\n"						\
+	"Content-Type: text/html; charset=iso-8859-1\r\n"		\
 	"Dummy7: 7\r\n"							\
 	"Dummy8: 8\r\n"							\
+	"Buzz: is hop-by-hop header\r\n"				\
+	"Dummy9: 9\r\n"							\
 	"Cache-Control: max-age=1, no-store, min-fresh=30\r\n"		\
 	"Pragma: no-cache, fooo \r\n"					\
 	"Cookie: session=42; theme=dark\r\n"				\
@@ -3096,8 +3094,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 16 total with 9 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
+		/* Common (raw) headers: 17 total with 10 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -3111,8 +3109,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 16 total with 9 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
+		/* Common (raw) headers: 17 total with 10 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -3134,8 +3132,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 16 total with 9 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
+		/* Common (raw) headers: 17 total with 10 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -3143,7 +3141,7 @@ TEST(http_parser, req_hop_by_hop)
 			case TFW_HTTP_HDR_CONNECTION:
 			case TFW_HTTP_HDR_KEEP_ALIVE:
 			case TFW_HTTP_HDR_RAW + 4:
-			case TFW_HTTP_HDR_RAW + 10:
+			case TFW_HTTP_HDR_RAW + 12:
 				EXPECT_TRUE(field->flags & TFW_STR_HBH_HDR);
 				break;
 			default:

--- a/fw/t/unit/test_http_parser.c
+++ b/fw/t/unit/test_http_parser.c
@@ -1659,11 +1659,7 @@ TEST(http_parser, eol_crlf)
 		EXPECT_TRUE(ht->tbl[TFW_HTTP_HDR_CONTENT_LENGTH].eolen == 1);
 	}
 
-	/*
-	 * It seems RFC 7230 3.3 doesn't prohibit message body
-	 * for GET requests.
-	 */
-	__FOR_REQ("GET / HTTP/1.1\n"
+	__FOR_REQ("POST / HTTP/1.1\n"
 		  "Host: b.com\n"
 		  "Content-Length: 6\n"
 		  "\r\n"
@@ -2914,12 +2910,9 @@ TEST(http_parser, req_hop_by_hop)
 	"Dummy5: 5\r\n"							\
 	"Foo: is hop-by-hop header\r\n"					\
 	"Dummy6: 6\r\n"							\
-	"Content-Length: 0\r\n"						\
-	"Content-Type: text/html; charset=iso-8859-1\r\n"		\
+	"Buzz: is hop-by-hop header\r\n"				\
 	"Dummy7: 7\r\n"							\
 	"Dummy8: 8\r\n"							\
-	"Buzz: is hop-by-hop header\r\n"				\
-	"Dummy9: 9\r\n"							\
 	"Cache-Control: max-age=1, no-store, min-fresh=30\r\n"		\
 	"Pragma: no-cache, fooo \r\n"					\
 	"Cookie: session=42; theme=dark\r\n"				\
@@ -2931,8 +2924,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 17 total with 10 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
+		/* Common (raw) headers: 16 total with 9 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -2946,8 +2939,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 17 total with 10 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
+		/* Common (raw) headers: 16 total with 9 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -2969,8 +2962,8 @@ TEST(http_parser, req_hop_by_hop)
 		REQ_HBH_END)
 	{
 		ht = req->h_tbl;
-		/* Common (raw) headers: 17 total with 10 dummies. */
-		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 17);
+		/* Common (raw) headers: 16 total with 9 dummies. */
+		EXPECT_EQ(ht->off, TFW_HTTP_HDR_RAW + 16);
 
 		for(id = 0; id < ht->off; ++id) {
 			field = &ht->tbl[id];
@@ -2978,7 +2971,7 @@ TEST(http_parser, req_hop_by_hop)
 			case TFW_HTTP_HDR_CONNECTION:
 			case TFW_HTTP_HDR_KEEP_ALIVE:
 			case TFW_HTTP_HDR_RAW + 4:
-			case TFW_HTTP_HDR_RAW + 12:
+			case TFW_HTTP_HDR_RAW + 10:
 				EXPECT_TRUE(field->flags & TFW_STR_HBH_HDR);
 				break;
 			default:

